### PR TITLE
Requesters containers

### PR DESCRIPTION
**Pull request by sstanculeanu**
## Reasoning behind the pull request
- new containers are needed for requesters
  
## Proposed changes
- add containers, factories and topic sender

## Testing procedure
- to be tested when feat branch is ready

**Note for review:** One who reviews should compare `requesterscontainer` package with `resolverscontainer` one, as the code was more or less moved. Didn't remove the logic from `resolverscontainer` yet in order to avoid enormous PRs, will be done in further ones. Same note for `requestersContainer.go`, which is almost the same with `resolversContainer.go`. Left a todo in order to generalize this container and reduce it to a single component used in both requesters and resolvers.

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/ElrondNetwork/elrond-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: dataRetriever/resolvers/topicResolverSender/topicResolverSender.go [hunk 1]
```diff
@@ -43,6 +43,7 @@ type ArgTopicResolverSender struct {
 	TargetShardId               uint32
 }
 
+// TODO[Sorin]: cleanup and move this into topicSender
```
Comment (iulianpascalau) [inline dataRetriever/resolvers/topicResolverSender/topicResolverSender.go:46]: :+1:

File: dataRetriever/topicSender/topicRequestSender_test.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+		NumIntraShardPeers:          2,
+		NumCrossShardPeers:          2,
+		NumFullHistoryPeers:         2,
+		CurrentNetworkEpochProvider: &mock.CurrentNetworkEpochProviderStub{},
+		SelfShardIdProvider:         mock.NewMultipleShardsCoordinatorMock(),
+		PeersRatingHandler:          &p2pmocks.PeersRatingHandlerStub{},
+	}
+}
+
+func TestNewTopicRequestSender(t *testing.T) {
```
Comment (iulianpascalau) [inline dataRetriever/topicSender/topicRequestSender_test.go:39]: I think we need more unit tests in this area. Can be another PR

File: dataRetriever/interface.go [hunk 3]
```diff
@@ -104,6 +113,32 @@ type ResolversContainerFactory interface {
...
+// RequestersContainer defines a requesters holder data type with basic functionality
+type RequestersContainer interface {
+	Get(key string) (Requester, error)
+	Add(key string, val Requester) error
+	AddMultiple(keys []string, requesters []Requester) error
+	Replace(key string, val Requester) error
+	Remove(key string)
+	Len() int
+	RequesterKeys() string
+	Iterate(handler func(key string, resolver Requester) bool)
```
Comment (ssd04) [inline dataRetriever/interface.go:137]:
```suggestion
	Iterate(handler func(key string, requester Requester) bool)
```
?

File: dataRetriever/topicSender/topicRequestSender_test.go (hunk 2 above)
Comment (sstanculeanu) [reply dataRetriever/topicSender/topicRequestSender_test.go:39]: added todo

File: dataRetriever/factory/requestersContainer/shardRequestersContainerFactory.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+	err = srcf.generateValidatorInfoRequester()
+	if err != nil {
+		return nil, err
+	}
+
+	return srcf.container, nil
+}
+
+// ------- Hdr requester
+
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply dataRetriever/factory/requestersContainer/shardRequestersContainerFactory.go:109]: deleted

File: dataRetriever/factory/requestersContainer/shardRequestersContainerFactory.go [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+		base.shardCoordinator.CommunicationIdentifier(base.shardCoordinator.SelfId())
+
+	return &shardRequestersContainerFactory{
+		baseRequestersContainerFactory: base,
+	}, nil
+}
+
+// Create returns a requesters container that will hold all requesters in the system
+func (srcf *shardRequestersContainerFactory) Create() (dataRetriever.RequestersContainer, error) {
+	err := srcf.generateTxRequesters(factory.TransactionTopic)
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply dataRetriever/factory/requestersContainer/shardRequestersContainerFactory.go]: pushed

File: dataRetriever/factory/containers/requestersContainer.go [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/core/container"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever"
+)
+
+var _ dataRetriever.RequestersContainer = (*requestersContainer)(nil)
+
+// requestersContainer is a requesters holder organized by type
+// TODO[Sorin]: next PR, add unittests and extract resolversContainer + requestersContainer in a general container struct
```
[the remark this answers is not in the document]
Comment (sstanculeanu) [reply dataRetriever/factory/containers/requestersContainer.go]: updated the todo

**Review by iulianpascalau — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]
